### PR TITLE
Add FormRectangularLinearSystem to ParMixedBilinearForm

### PR DESCRIPTION
**Pull request by jandrej**
This PR adds FormRectangularLinearSystem to the parallel classes. Those extend the serial version in #1164.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1259](https://github.com/mfem/mfem/pull/1259) | @jandrej | @v-dobrev | @pazner + @tzanio | 01/28/20 | 02/19/20 | 02/23/20 | |

### CODE REVIEW

**Comment by tzanio** [comment]
@v-dobrev, can you take this one please

**Comment by jandrej** [comment]
Ping @v-dobrev @tzanio

**Comment by pazner** [comment]
It looks good to me. Is there a simple example I can easily test out?

**Comment by jandrej** [comment]
Should we put a stokes like system with this in the tests or extend the darcy flow example?

**Comment by pazner** [comment]
I think Stokes example would be great, but is probably beyond the scope of this PR. Extending Darcy could be a good option. Or just adding a small test and verifying e.g. that BCs are enforced properly.

**Comment by jandrej** [comment]
The last commit changes the unit tests system s.t. it can be run in parallel.

File: tests/unit/unit_test_main.cpp [hunk 1]
```diff
@@ -9,5 +9,17 @@
...
-#define CATCH_CONFIG_MAIN     // This tells Catch to provide a main() - only do this in one cpp file
+#define CATCH_CONFIG_RUNNER
 #include "catch.hpp"
+#include "mfem.hpp"
+
+int main(int argc, char *argv[])
+{
+#ifdef MFEM_USE_MPI
+   mfem::MPI_Session mpi;
+#endif
```
Comment (pazner) [inline tests/unit/unit_test_main.cpp]: Note that some of the tests (e.g. the data collection tests that do file I/O) will fail if you run in parallel. Maybe we should force only one MPI rank?

File: fem/pbilinearform.hpp [hunk 2]
```diff
@@ -241,6 +246,15 @@ class ParMixedBilinearForm : public MixedBilinearForm
...
    void ParallelAssemble(OperatorHandle &A);
 
+   virtual void FormRectangularSystemMatrix(const Array<int> &trial_tdof_list,
+                                            const Array<int> &test_tdof_list,
+                                            OperatorHandle &A);
+
+   virtual void FormRectangularLinearSystem(const Array<int> &trial_tdof_list,
+                                            const Array<int> &test_tdof_list, Vector &x,
+                                            Vector &b, OperatorHandle &A, Vector &X,
+                                            Vector &B);
```
Comment (pazner) [inline fem/pbilinearform.hpp]: Documentation would be helpful here. Even just a reference to look at the serial versions of these.

File: tests/unit/unit_test_main.cpp (hunk 1 above)
Comment (pazner) [reply tests/unit/unit_test_main.cpp]: I think this is the only thing still outstanding from my review. Ideas?

Comment (jandrej) [reply tests/unit/unit_test_main.cpp]: I missed that. If those fail, my suggestion is to not run them in parallel if they are serial only. E.g.

if compiled with MPI -> if num_ranks > 1 -> skip this test by returning a success

Comment (tzanio) [reply tests/unit/unit_test_main.cpp]: If I understand correctly, this is mostly about compilation -- the unit tests are actually run without `mpirun` so there shouldn't be any issues with the current tests.

File: tests/unit/unit_test_main.cpp [hunk 3]
```diff
@@ -9,5 +9,40 @@
...
+   {
+      auto cfg = session.configData();
+      cfg.testsOrTags.push_back("[Parallel]");
+      session.useConfigData(cfg);
+   }
+   if (mpi.WorldSize() > 1 && mpi.Root())
+   {
+      mfem::out
+         << "WARNING: Only running the [Parallel] label on MPI ranks > 1."
+         << std::endl;
```
Comment (pazner) [inline tests/unit/unit_test_main.cpp]: `make style`

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Can you take a look at [this](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/e4b69b607185c41eca91d2e72be812ae57bd02f2#tux426/2020-02-21-llnl-gh-next/unit-test.out) regression test error:

```
g++ -g -c /home/kolev1/autotest-mfem/tux426/mfem/tests/unit/miniapps/test_sedov.cpp -I/home/kolev1/autotest-mfem/tux426/mfem/tests/unit -I/home/kolev1/autotest-mfem/tux426/mfem  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -DMFEM_SEDOV_TESTS -DMFEM_SEDOV_DEVICE='"debug"' -o miniapps/test_sedov.debug.o
g++ -g -c /home/kolev1/autotest-mfem/tux426/mfem/tests/unit/miniapps/test_sedov.cpp -I/home/kolev1/autotest-mfem/tux426/mfem/tests/unit -I/home/kolev1/autotest-mfem/tux426/mfem  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -DMFEM_SEDOV_TESTS -DMFEM_SEDOV_DEVICE='"cpu"' -o miniapps/test_sedov.cpu.o
/home/kolev1/autotest-mfem/tux426/mfem/tests/unit/unit_test_main.cpp:13:20: fatal error: mfem.hpp: No such file or directory
 #include "mfem.hpp"
                    ^
compilation terminated.
make: *** [/home/kolev1/autotest-mfem/tux426/mfem/tests/unit/unit_test_main.o] Error 1
```

**Comment by jandrej** [comment]
Works for me.

```
> echo $PWD
/Users/andrej1/repos/mfem-merge/tests/unit

> make clean
rm -f unit_tests  *.o */*.o */*~ *~
rm -rf *.dSYM output_meshes

> make
mpicxx -g -c /Users/andrej1/repos/mfem-merge/tests/unit/unit_test_main.cpp -I. -I../..  -g -Wall -std=c++11 -I../.. -I/Users/andrej1/local/hypre-2.15.1/include -I/Users/andrej1/local/parmetis-4.0.3/include -g -o unit_test_main.o
....
mpicxx -g unit_test_main.o fem/test_1d_bilininteg.o fem/test_2d_bilininteg.o fem/test_3d_bilininteg.o fem/test_assemblediagonalpa.o fem/test_calcshape.o fem/test_datacollection.o fem/test_fe.o fem/test_intrules.o fem/test_intruletypes.o fem/test_inversetransform.o fem/test_lin_interp.o fem/test_linear_fes.o fem/test_operatorjacobismoother.o fem/test_pa_coeff.o fem/test_pa_kernels.o fem/test_quadraturefunc.o general/text-test.o linalg/test_blockMatrix.o linalg/test_complex_operator.o linalg/test_densematrix.o linalg/test_ilu.o linalg/test_ode.o linalg/test_operator.o linalg/test_rectangularmatrix.o mesh/test_mesh.o -I. -I../.. -g -Wall -std=c++11 -I../.. -I/Users/andrej1/local/hypre-2.15.1/include -I/Users/andrej1/local/parmetis-4.0.3/include -g -Wl,-undefined,dynamic_lookup -L../.. -lmfem -Wl,-rpath,/Users/andrej1/local/hypre-2.15.1/lib -L/Users/andrej1/local/hypre-2.15.1/lib -lHYPRE  -Wl,-rpath,/Users/andrej1/local/parmetis-4.0.3/lib -L/Users/andrej1/local/parmetis-4.0.3/lib -lmetis -o unit_tests
```

**Comment by camierjs** [comment]
Couldn't reproduce it neither.

**Comment by tzanio** [comment]
This is actually a conflict between this PR and #1104 -- they both introduce MPI communicators in different places, so both need to change for this to be fixed.

We will first make the change here, see https://github.com/mfem/mfem/pull/1259/commits/61fe962cd269e0175cc284181f1a6d3c959d76be.

Then, after this PR is merged and #1104 is updated, the `test_sedov` test will need to be updated to use `GlobalMPISession`.

**Comment by jandrej** [comment]
Re-merged in `next` for testing...

**Pull request merged**